### PR TITLE
Replace recursions with iterations in the linked_list data structure

### PR DESCRIPTION
**Pull request by rtakacs**
I measured the stack usage of IoT.js on an ARM board. On the peak of the stack usage I created a backtrace (with stack frame sizes) to see which part of code uses significant memory.

The test cases were an empty file, and another with only the 'http' module included. The following backtrace is related to the second one (but the first one is also similar):

```
#0	__cyg_profile_func_exit	32 byte
#1	mem_get_block_chunks_count_from_data_size	16 byte
#2	mem_heap_alloc_block_internal	32 byte
#3	mem_heap_alloc_block_try_give_memory_back	24 byte
#4	mem_heap_alloc_block	16 byte
#5	jsp_mm_alloc	24 byte
#6	linked_list_init	24 byte
#7	linked_list_set_element	32 byte
#8	linked_list_set_element	32 byte
...
#366	linked_list_set_element	32 byte
#367	scopes_tree_add_op_meta	40 byte
...
#501	main	32 byte
```

The first '...' means further 356 linked_list_set_element calls. The second '...' is not interesting, just hidden to make the backtrace smaller.

You can see that <b>linked_list_set_element</b> is in recursion 359 times.

In this patch I rewrite functions to use iterations, and the result is shown by the following table:

Test cases | recursion | iteration | diff
------------ | -------------|------------|------
empty file | 11.52 K | 10.71 K | <b>0.81 K</b>
require('http') | 19.51K | 14.12 K | <b>5.39 K</b>

Also need to mention the runtime of this test cases:

Test cases | recursion | iteration | gain
------------ | -------------|------------|------
empty file |  0.192 s | 0.153 | <b>20.3125 %</b>
require('http') | 0.755 s | 0.648 s | <b>14.1722 %</b>

<b>Checked IoT.js revision:</b> f3bee70
<b>Contained Jerry revision: </b>a00079e
<b>Build type:</b> Release

### CODE REVIEW

File: jerry-core/parser/js/collections/linked-list.cpp [hunk 1]
```diff
@@ -57,11 +57,13 @@ linked_list_free (linked_list list)
 {
   ASSERT_LIST (list);
   linked_list_header *header = (linked_list_header *) list;
-  if (header->next)
+
+  while (header != null_list)
   {
-    linked_list_free ((linked_list) header->next);
+    linked_list_header *next_header = header->next;
```
Comment (egavrin) [inline jerry-core/parser/js/collections/linked-list.cpp]: `next_header_p`

File: jerry-core/parser/js/collections/linked-list.cpp [hunk 2]
```diff
@@ -69,45 +71,53 @@ linked_list_element (linked_list list, size_t element_num)
...
     {
       return NULL;
     }
+
+    element_count = element_count - (block_size / header->element_size);
   }
-  raw += header->element_size * element_num;
-  return raw;
+
+  uint8_t *raw_start = (linked_list) header + sizeof (linked_list_header);
```
Comment (egavrin) [inline jerry-core/parser/js/collections/linked-list.cpp]: `raw_start_p`

File: jerry-core/parser/js/collections/linked-list.cpp [hunk 3]
```diff
@@ -69,45 +71,53 @@ linked_list_element (linked_list list, size_t element_num)
...
-  if (element == NULL)
-  {
-    return;
+
+    header = header->next;
+    element_count = element_count - (block_size / header->element_size);
   }
-  memcpy (raw + element_num * header->element_size, element, header->element_size);
+
+  uint8_t *raw_start = (linked_list) header + sizeof (linked_list_header);
```
Comment (egavrin) [inline jerry-core/parser/js/collections/linked-list.cpp]: `raw_start_p`

**Comment by egavrin** [comment]
Benchmark |         RSS<br>(+ is better) |        Perf<br>(+ is better) |
                               --------- |                          --- |                         ---- |
                              3d-cube.js |          132->   132 (0.000) |      3.69867-> 3.684 (0.397) | 
                  access-binary-trees.js |           88->    88 (0.000) |    2.73733->2.74267 (-0.195) | 
                      access-fannkuch.js |           40->    40 (0.000) |    9.86533->9.91867 (-0.541) | 
                         access-nbody.js |           64->    64 (0.000) |     4.63067->4.62667 (0.086) | 
             bitops-3bit-bits-in-byte.js |           36->    36 (0.000) |       3.292->3.24133 (1.539) | 
                  bitops-bits-in-byte.js |           36->    36 (0.000) |    4.41867->4.44933 (-0.694) | 
                   bitops-bitwise-and.js |           32->    32 (0.000) |    4.12933->4.14533 (-0.387) | 
                controlflow-recursive.js |          224->   224 (0.000) |    3.19467->3.20267 (-0.250) | 
                    date-format-xparb.js |          100->   100 (0.000) |      2.39067-> 2.352 (1.617) | 
                          math-cordic.js |           48->    48 (0.000) |     4.84133->4.82267 (0.385) | 
                    math-partial-sums.js |           40->    40 (0.000) |        2.564-> 2.564 (0.000) | 
                   math-spectral-norm.js |           52->    52 (0.000) |    3.19733->3.23867 (-1.293) | 
                         string-fasta.js |           56->    56 (0.000) |     15.0107->14.9987 (0.080) | 
                         Geometric mean: |            RSS reduction: 0% |            Speed up: 0.0604% |

**Comment by egavrin** [comment]
Looks good to me. Any change that reduces stack usage is good for me.

Let @sand1k to check it too.

**Comment by zherczeg** [comment]
LGTM

**Comment by sand1k** [comment]
LGTM.

**Comment by galpeter** [comment]
merged: 6f7746050984787b5e578bbdb17a70c81089f696

**Comment by muddog** [comment]
Hi All

How can I run the sunspider benchmark tool on the mbedOS with JerryScript?
Anybody can help?

Thanks.

**Pull request closed without merging**